File: 993/find_period.py
import sys

lines = open('bb_values.txt').read().splitlines()
bbs = {}
for line in lines:
    if line.strip():
        parts = line.split()
        if len(parts) >= 2:
            try:
                bbs[int(parts[0])] = int(parts[1])
            except:
                pass

N_vals = sorted(bbs.keys())
V = [bbs[n] for n in N_vals]

print(f"Loaded {len(V)} values.")

def find_period(seq, min_p, max_p):
    for p in range(min_p, max_p+1):
        # check if sequence ultimately becomes periodic with period p
        # differences seq[i+p] - seq[i] should be constant
        valid = False
        for start in range(len(seq) - 200*p):
            # Try to see if it becomes constant after `start`
            constant = True
            diff = seq[start+p] - seq[start]
            for i in range(start, len(seq) - p):
                if seq[i+p] - seq[i] != diff:
                    constant = False
                    break
            if constant:
                print(f"Found period P={p}, shift={start}, diff={diff}")
                return p, start, diff
    return None

find_period(V, 1, 1000)

File: 993/find_period.c
#include <stdio.h>
#include <stdlib.h>

long long V[200000];
int V_len = 0;

int main() {
    FILE *f = fopen("bb_values.txt", "r");
    if (!f) return 1;
    long long n, v;
    while (fscanf(f, "%lld %lld", &n, &v) == 2) {
        V[V_len++] = v;
    }
    fclose(f);
    
    printf("Loaded %d values.\n", V_len);
    
    for (int p = 1; p <= 5000; p++) {
        for (int start = 0; start < V_len - 100 * p; start++) {
            int constant = 1;
            long long diff = V[start + p] - V[start];
            for (int i = start; i < V_len - p; i++) {
                if (V[i + p] - V[i] != diff) {
                    constant = 0;
                    break;
                }
            }
            if (constant) {
                printf("Found period P=%d, shift=%d, diff=%lld\n", p, start, diff);
                return 0;
            }
        }
    }
    printf("No period found up to P=5000\n");
    return 0;
}
